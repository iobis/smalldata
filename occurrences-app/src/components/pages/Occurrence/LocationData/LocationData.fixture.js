import LocationData from './LocationData'

export default [{
  component: LocationData,
  name:      'for empty data',
  props:     {
    data:     getEmptyData(),
    onChange: console.log
  }
}]

export function getEmptyData() {
  return {
    decimalLongitude:      null,
    decimalLatitude:       null,
    coordinateUncertainty: null,
    minimumDepth:          null,
    maximumDepth:          null,
    verbatimCoordinates:   '',
    verbatimEventDate:     '',
    verbatimDepth:         ''
  }
}
