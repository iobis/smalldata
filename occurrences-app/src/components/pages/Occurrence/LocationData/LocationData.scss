.location-data {
  margin-bottom: 1.25rem;

  .event-begin-date {
    .datetimepicker-clear-button {
      display: none;
    }
  }

  .verbatim-data {
    background-color: $background;
    padding: 1.25em 1.5em;
    border-radius: $radius;
  }

  .location-picker-notice {
    background-color: var(--obis-green);
    cursor: pointer;
    color: #fff;
    margin-bottom: var(--margin-bottom);
  }

  .no-margin {
    margin-bottom: 0;
  }
}

.modal-card-head {
  margin-bottom: 0;
}
