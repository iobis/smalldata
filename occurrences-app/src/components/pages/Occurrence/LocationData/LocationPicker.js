import 'leaflet/dist/leaflet.css'
import iconRetinaUrl from 'leaflet/dist/images/marker-icon-2x.png'
import iconUrl from 'leaflet/dist/images/marker-icon.png'
import L from 'leaflet'
import PropTypes from 'prop-types'
import React, { useState, useEffect, useRef } from 'react'
import shadowUrlUrl from 'leaflet/dist/images/marker-shadow.png'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { Map, TileLayer, Marker, Popup } from 'react-leaflet'
import { useDebounce } from '../../../../hooks/hooks'
import { useTranslation } from 'react-i18next'

// This part of code relates to issue with react-leaflet. Please update the code when issues is fixed. Refer
// to https://github.com/PaulLeCam/react-leaflet/issues/453 for more details
const DefaultIcon = L.icon({
  iconRetinaUrl: iconRetinaUrl,
  iconUrl:       iconUrl,
  shadowUrl:     shadowUrlUrl
})

L.Marker.prototype.options.icon = DefaultIcon

export default function LocationPicker({ onChange }) {
  const { t } = useTranslation()
  const mapRef = useRef()
  const [latitude, setLatitude] = useState(51.505)
  const [longitude, setLongitude] = useState(-0.09)
  const coordinates = [latitude, longitude]
  const [zoom, setZoom] = useState(12)
  const [searchString, setSearchString] = useState('')
  const [suggestions, setSuggestions] = useState([])

  function setMarkerCoordinates(latlng) {
    setLatitude(latlng.lat)
    setLongitude(latlng.lng)
    onChange({ latitude: latlng.lat, longitude: latlng.lng })
  }

  const debouncedSearch = useDebounce(searchString, 500)

  useEffect(() => {
    const fetchSuggestions = async() => {
      const response = await fetch(`https://nominatim.openstreetmap.org/search?format=json&q=${searchString}`)
      const suggestions = await response.json()
      setSuggestions(suggestions)
    }

    if (debouncedSearch) fetchSuggestions()
  }, [debouncedSearch])

  useEffect(() => {
    mapRef.current.leafletElement.invalidateSize(true)
  })

  function handleClearClick() {
    setSuggestions([])
    setSearchString('')
  }

  function handleSearchStringChange(newSearchString) {
    setSearchString(newSearchString)
    if (!newSearchString) setSuggestions([])
  }

  return (
    <div className="location-picker section is-fluid">
      <div className="columns">
        <div className="column is-half">
          <div className="location-search field">
            <div className="control has-icons-left has-icons-right">
              <input
                className="input"
                onChange={(e) => handleSearchStringChange(e.target.value)}
                placeholder={t('occurrenceForm.locationData.locationPicker.searchTip')}
                type="text"
                value={searchString}/>
              <span className="icon is-left">
                <FontAwesomeIcon className="search" icon="search"/>
              </span>
              <span className="clear icon is-small is-right">
                <FontAwesomeIcon className="times-circle" icon="times-circle" onClick={handleClearClick}/>
              </span>
            </div>
          </div>
          <h4 className="result title is-4">{t('occurrenceForm.locationData.locationPicker.resultsTitle')}</h4>
          {suggestions.length > 0
            ? <SuggestionsResult onClick={setMarkerCoordinates} suggestions={suggestions}/>
            : <SuggestionsResultEmpty/>}
        </div>
        <div className="column is-half">
          <Map
            center={coordinates}
            onClick={(e) => setMarkerCoordinates(e.latlng)}
            onZoomEnd={(e) => setZoom(e.target.getZoom())}
            ref={mapRef}
            zoom={zoom}>
            <TileLayer
              attribution="&amp;copy <a href=&quot;http://osm.org/copyright&quot;>OpenStreetMap</a> contributors"
              url="https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png"/>
            <Marker
              draggable={true}
              onDragend={(e) => setMarkerCoordinates(e.target.getLatLng())}
              position={coordinates}>
              <Popup>
                <div>latitude: {latitude}</div>
                <div>longitude: {longitude}</div>
              </Popup>
            </Marker>
          </Map>
        </div>
      </div>
    </div>
  )
}

LocationPicker.propTypes = {
  onChange: PropTypes.func.isRequired
}

function SuggestionsResult({ suggestions, onClick }) {
  const { t } = useTranslation()

  return (
    <table className="table is-fullwidth is-striped is-hoverable">
      <thead>
      <tr>
        <th>{t('common.type')}</th>
        <th>{t('common.name')}</th>
        <th>{t('common.longitude')}</th>
        <th>{t('common.latitude')}</th>
      </tr>
      </thead>
      <tbody>
      {suggestions.map(suggestion => (
        <tr
          className="suggestion-row fieldrow"
          key={suggestion.place_id}
          onClick={() => onClick({ lat: suggestion.lat, lng: suggestion.lon })}>
          <td className="type">
            {suggestion.type}
          </td>
          <td className="name">
            {suggestion.display_name}
          </td>
          <td className="longitude">
            {suggestion.lon}
          </td>
          <td className="latitude">
            {suggestion.lat}
          </td>
        </tr>
      ))}
      </tbody>
    </table>
  )
}

SuggestionsResult.propTypes = {
  onClick:     PropTypes.func.isRequired,
  suggestions: PropTypes.array.isRequired
}

function SuggestionsResultEmpty() {
  const { t } = useTranslation()

  return (
    <>
      <p>{t('occurrenceForm.locationData.locationPicker.emptyResult')}</p>
      <p>{t('occurrenceForm.locationData.locationPicker.emptyResultTip')}</p>
    </>
  )
}
