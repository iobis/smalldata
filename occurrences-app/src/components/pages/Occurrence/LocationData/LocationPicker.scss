.location-picker {
  .result.title {
    margin-bottom: 0.5rem;
  }

  .times-circle {
    cursor: pointer;
    pointer-events: initial;
  }

  .suggestion-row {
    cursor: pointer;
  }

  .leaflet-container {
    height: 500px;
    width: 100%;
  }
}
