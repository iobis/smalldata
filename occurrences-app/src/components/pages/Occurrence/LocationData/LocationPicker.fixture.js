import LocationPicker from './LocationPicker'

export default [{
  component: LocationPicker,
  name:      'default',
  props:     {
    onChange: console.log
  }
}]
