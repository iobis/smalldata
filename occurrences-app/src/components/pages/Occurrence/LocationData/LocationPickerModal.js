import classNames from 'classnames'
import LocationPicker from './LocationPicker'
import PropTypes from 'prop-types'
import React, { useState } from 'react'
import { useTranslation } from 'react-i18next'

export default function LocationPickerModal({ active, onClose, onChange }) {
  const { t } = useTranslation()
  const [location, setLocation] = useState({ latitude: 0, longitude: 0 })

  return (
    <div className={classNames('location-picker-modal modal', { 'is-active': active })}>
      <div className="modal-background"/>
      <div className="modal-card">
        <header className="modal-card-head">
          <p className="modal-card-title">{t('occurrenceForm.locationData.locationPicker.title')}</p>
          <button aria-label="close" className="delete" onClick={() => onClose()}/>
        </header>
        <section className="modal-card-body">
          <LocationPicker onChange={(location) => setLocation(location)}/>
        </section>
        <footer className="modal-card-foot">
          <button className="button is-success" onClick={() => onChange(location)}>{t('common.confirm')}</button>
          <button className="button" onClick={() => onClose()}>{t('common.cancel')}</button>
        </footer>
      </div>
    </div>
  )
}

LocationPickerModal.propTypes = {
  active:   PropTypes.bool.isRequired,
  onChange: PropTypes.func.isRequired,
  onClose:  PropTypes.func.isRequired
}
