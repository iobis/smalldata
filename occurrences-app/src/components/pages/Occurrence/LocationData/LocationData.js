import CopyPreviousData from '../CopyPreviousData'
import InputNumber from '../../../form/InputNumber'
import InputText from '../../../form/InputText'
import LocationPickerModal from './LocationPickerModal'
import PropTypes from 'prop-types'
import React, { useState } from 'react'
import { useTranslation } from 'react-i18next'
import ContinueButton from '../ContinueButton'

export default function LocationData({ data, onChange, nextStepHandler }) {
  const { t } = useTranslation()
  const updateField = (name, value) => {
    const newSelection = { ...data, [name]: value }
    onChange(newSelection)
  }
  const [locationPickerVisible, setLocationPickerVisible] = useState(false)

  function handleLocationUpdate(location) {
    setLocationPickerVisible(false)
    const newSelection = { ...data, decimalLatitude: location.latitude, decimalLongitude: location.longitude }
    onChange(newSelection)
  }

  return (
    <div className="location-data section is-fluid">
      <h1 className="title">{t('occurrenceForm.locationData.enterCoordinates.title')}</h1>
      <h2 className="subtitle">{t('occurrenceForm.locationData.enterCoordinates.subtitle')}</h2>
      <div className="columns">
        <InputNumber
          className="decimal-longitude is-3"
          name="occurrenceForm.locationData.decimalLongitude"
          onChange={(value) => updateField('decimalLongitude', value)}
          step={0.0001}
          value={data.decimalLongitude}/>
        <InputNumber
          className="decimal-latitude is-3"
          name="occurrenceForm.locationData.decimalLatitude"
          onChange={(value) => updateField('decimalLatitude', value)}
          step={0.0001}
          value={data.decimalLatitude}/>
        <InputNumber
          className="coordinate-uncertainty is-3"
          name="occurrenceForm.locationData.coordinateUncertainty"
          onChange={(value) => updateField('coordinateUncertainty', value)}
          optional
          step={0.01}
          value={data.coordinateUncertainty}/>
      </div>
      <div className="columns">
        <InputNumber
          className="minimum-depth is-3"
          name="occurrenceForm.locationData.minimumDepth"
          onChange={(value) => updateField('minimumDepth', value)}
          optional
          step={0.01}
          value={data.minimumDepth}/>
        <InputNumber
          className="maximum-depth is-3"
          name="occurrenceForm.locationData.maximumDepth"
          onChange={(value) => updateField('maximumDepth', value)}
          optional
          step={0.01}
          value={data.maximumDepth}/>
      </div>
      <div className="columns">
        <div className="location-picker-notice column is-narrow" onClick={() => setLocationPickerVisible(true)}>
          {t('occurrenceForm.locationData.locationPicker.title')}
        </div>
      </div>
      <LocationPickerModal
        active={locationPickerVisible}
        onChange={handleLocationUpdate}
        onClose={() => setLocationPickerVisible(false)}/>
      <div className="verbatim-data">
        <h1 className="title">{t('occurrenceForm.locationData.verbatimData.title')}</h1>
        <h2 className="subtitle">{t('occurrenceForm.locationData.verbatimData.subtitle')}</h2>
        <div className="columns">
          <InputText
            className="verbatim-coordinates is-3"
            name="occurrenceForm.locationData.verbatimCoordinates"
            onChange={(value) => updateField('verbatimCoordinates', value)}
            optional
            value={data.verbatimCoordinates}/>
          <InputText
            className="verbatim-depth is-3"
            name="occurrenceForm.locationData.verbatimDepth"
            onChange={(value) => updateField('verbatimDepth', value)}
            optional
            value={data.verbatimDepth}/>
        </div>
      </div>
      <CopyPreviousData/>
      <ContinueButton
        name="datasetContinue"
        nextStepHandler={nextStepHandler}
        value="occurrenceForm.locationData.step.nextStep"
        wrapperClassName=""/>
    </div>
  )
}

export const locationDataShape = {
  decimalLongitude:      PropTypes.number,
  decimalLatitude:       PropTypes.number,
  coordinateUncertainty: PropTypes.number,
  minimumDepth:          PropTypes.number,
  maximumDepth:          PropTypes.number,
  verbatimCoordinates:   PropTypes.string.isRequired,
  verbatimDepth:         PropTypes.string.isRequired
}

LocationData.propTypes = {
  data:     PropTypes.shape(locationDataShape).isRequired,
  nextStepHandler: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired
}
