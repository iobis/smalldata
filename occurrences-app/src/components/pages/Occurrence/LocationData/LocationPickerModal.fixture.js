import LocationPickerModal from './LocationPickerModal'
import { faSearch, faTimesCircle } from '@fortawesome/free-solid-svg-icons'
import { library } from '@fortawesome/fontawesome-svg-core'

library.add(faSearch, faTimesCircle)

export default [{
  component: LocationPickerModal,
  name:      'default',
  props:     {
    active:   true,
    onChange: console.log,
    onClose:  console.log
  }
}]
