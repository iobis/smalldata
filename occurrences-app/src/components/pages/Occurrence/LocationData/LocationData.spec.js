import React from 'react'
import LocationData from './LocationData'
import { mount } from 'enzyme'

describe('SelectDataset', () => {
  it('renders correctly', () => {
    expect(mount(createComponent())).toMatchSnapshot()
  })
})

function createComponent() {
  const data = {
    decimalLongitude:      '',
    decimalLatitude:       '',
    coordinateUncertainty: '',
    minimumDepth:          '',
    maximumDepth:          ''
  }
  const defaultProps = { data, onChange: () => {} }
  return <LocationData {...defaultProps}/>
}
