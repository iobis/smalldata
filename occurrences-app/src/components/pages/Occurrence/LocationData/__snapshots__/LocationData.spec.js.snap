// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectDataset renders correctly 1`] = `
<LocationData
  data={
    Object {
      "coordinateUncertainty": "",
      "decimalLatitude": "",
      "decimalLongitude": "",
      "maximumDepth": "",
      "minimumDepth": "",
    }
  }
  onChange={[Function]}
>
  <div
    className="location-data section is-fluid"
  >
    <div
      className="columns"
    >
      <InputText
        className="decimal-longitude is-2"
        name="occurrenceForm.locationData.decimalLongitude"
        onChange={[Function]}
        value=""
      >
        <div
          className="column field decimal-longitude is-2"
        >
          <label
            className="label"
          >
            occurrenceForm.locationData.decimalLongitude.label
          </label>
          <input
            className="input"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
      </InputText>
      <InputText
        className="decimal-latitude is-2"
        name="occurrenceForm.locationData.decimalLatitude"
        onChange={[Function]}
        value=""
      >
        <div
          className="column field decimal-latitude is-2"
        >
          <label
            className="label"
          >
            occurrenceForm.locationData.decimalLatitude.label
          </label>
          <input
            className="input"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
      </InputText>
      <InputText
        className="coordinate-uncertainty is-2"
        name="occurrenceForm.locationData.coordinateUncertainty"
        onChange={[Function]}
      >
        <div
          className="column field coordinate-uncertainty is-2"
        >
          <label
            className="label"
          >
            occurrenceForm.locationData.coordinateUncertainty.label
          </label>
          <input
            className="input"
            onChange={[Function]}
            type="text"
          />
        </div>
      </InputText>
    </div>
    <div
      className="columns"
    >
      <InputText
        className="minimum-depth is-2"
        name="occurrenceForm.locationData.minimumDepth"
        onChange={[Function]}
        value=""
      >
        <div
          className="column field minimum-depth is-2"
        >
          <label
            className="label"
          >
            occurrenceForm.locationData.minimumDepth.label
          </label>
          <input
            className="input"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
      </InputText>
      <InputText
        className="maximum-depth is-2"
        name="occurrenceForm.locationData.maximumDepth"
        onChange={[Function]}
        value=""
      >
        <div
          className="column field maximum-depth is-2"
        >
          <label
            className="label"
          >
            occurrenceForm.locationData.maximumDepth.label
          </label>
          <input
            className="input"
            onChange={[Function]}
            type="text"
            value=""
          />
        </div>
      </InputText>
    </div>
  </div>
</LocationData>
`;
